### ISSUE TITLE
[WIP] Support SSR

### ISSUE THREAD

**Issue by elitan** [nhost/hasura-backend-plus#280]
Right now HBP supports web [SPA](https://en.wikipedia.org/wiki/Single-page_application)s (ex Create React App) really well with its httpOnly cookie strategy for refresh token. However, this strategy does not play well with SSR (ex some parts of NextJS).

Even if [SSR is getting less and less use cases](https://arunoda.me/blog/hey-nextjs-is-server-side-rendering-dead), there still are some cases where SSR is needed.

HBP should work with SSR. And currently HBP does work with SSR **if** HBP and the frontend both run on the same domain becuase then the refresh_token gets shared between frontend and backend.

But if your backend and frontend is not sharing domain, they wont share cookies either. So if your backend is on `backend-xxx.nhost.app` you can not use HBP and SSR if your frontend is on `localhost` or `example.com`.

I propose HBP adds a flag to send the refresh token in the http body so the frontend can set the refresh token as a cookie. This means that `/auth/token/refresh?ssr=true` and `/auth/login?ssr=true` should send two refresh_tokens. One as usual as a cookie and one in the http body.

The cookie in the body should be set by the frontend

**Comment by elitan** [comment]
Also, what are some use cases when SSR (under auth) is neccessary? Why not do this:

| Side   	| Auth needed 	| TODO                                                  	|
|--------	|-------------	|-------------------------------------------------------	|
| Server 	| True        	| Why use SSR? Why not fetch the data from the client?  	|
| Server 	| False       	| Use `HASURA_GRAPHQL_UNAUTHORIZED_ROLE` to access data 	|
| Client 	| True        	| Use `nhost-js-sdk` as usual.                          	|
| Client 	| False       	| Use `HASURA_GRAPHQL_UNAUTHORIZED_ROLE` to access data 	|

**Comment by alveshelio** [comment]
Hi @elitan,

I've read the article and there were some points I wasn't aware of. 

Same for the role of `HASURA_GRAPHQL_UNAUTHORIZED_ROLE`, I didn't know about it and I'll need to dig a bit more and see where and how to use it.

**Comment by defrex** [comment]
I agree that Server-Side Rendering has become less critical as more complex Static Rendering scenarios have gained support (eg Gatsby or Next.js's `getStaticProps`). When auth is not required (or only a minor portion of the page is affected by auth) Static Rendering (potentially with "auth sprinkles") is likely preferred.

However, in my opinion, Single Page Apps are worse then Server-Side Rendered apps for virtually every other auth-bound use-case. Server rendering gives you the load-time performance characteristics of rails-style dynamic websites with the run-time characteristics of a single-page app. This "best of both worlds" performance is a massive win, IMO. "Time to paint" gets driven way down and makes the subjective feeling of responsiveness much stronger.

Speaking for myself, not having SSR support is a deal-breaker for a platform like Nhost.

That said, putting Hasura on the same domain as the client code is perfectly reasonable as the only way to make it work. If that does the job without additional features, I'd stop there.

**Comment by ConProgramming** [comment]
> I agree that Server-Side Rendering has become less critical as more complex Static Rendering scenarios have gained support (eg Gatsby or Next.js's `getStaticProps`). When auth is not required (or only a minor portion of the page is affected by auth) Static Rendering (potentially with "auth sprinkles") is likely preferred.
> 
> However, in my opinion, Single Page Apps are worse then Server-Side Rendered apps for virtually every other auth-bound use-case. Server rendering gives you the load-time performance characteristics of rails-style dynamic websites with the run-time characteristics of a single-page app. This "best of both worlds" performance is a massive win, IMO. "Time to paint" gets driven way down and makes the subjective feeling of responsiveness much stronger.
> 
> Speaking for myself, not having SSR support is a deal-breaker for a platform like Nhost.
> 
> That said, putting Hasura on the same domain as the client code is perfectly reasonable as the only way to make it work. If that does the job without additional features, I'd stop there.

Yep, I agree SSR support is a big deal for me. I use https://sapper.svelte.dev/ in most projects which uses SSR and I'm disappointed to find out I'll have to change our serverside auth (which works with Nhost rn except for refresh tokens) to client side just to get refresh tokens to work.
